pyproject.toml

```diff
@@ -46,27 +46,34 @@ Documentation = "https://jupytext.readthedocs.io"
 
 [project.optional-dependencies]
 # Test related dependencies
-# TODO: Split them into unit, integration and functional tests groups
 test = [
-    "autopep8",
-    "black",
-    "isort",
-    "ruff",
-    "flake8",
-    "pytest",
-    "pytest-randomly",
-    "gitpython",
-    "jupyterlab",
-    "notebook",
-    "nbconvert",
-    # jupyter-fs==0.4.0 is async, which is not supported by Jupytext ATM
-    "jupyter-fs<0.4.0",
-    "ipykernel",
-    "pre-commit",
+  "pytest",
+  "pytest-randomly"
+]
+test-integration = [
+  "jupytext[test]",
+  "jupyter-server",
+  "nbconvert"
+]
+test-functional = [
+  "jupytext[test-integration]",
+  # jupytext --pipe and --check
+  "autopep8",
+  "black",
+  "isort",
+  "flake8",
+  # jupytext --execute
+  "ipykernel",
+  # Pre-commit tests
+  "gitpython",
+  "pre-commit",
+  # Interaction with other contents managers
+  # jupyter-fs==0.4.0 is async, which is not supported by Jupytext ATM
+  "jupyter-fs<0.4.0"
```

Comment:
I would prefer thse in an integration test, mostly because these are optional dependencies, and they primarily test how `jupytext` interacts with other tools. Indeed the tests are better classified as functional tests, and I see many projects do so, so it is more of a subjective classification.

Why I would prefer this as `integration` is because w.r.t. packaging, unit and functional tests are good to include as a whole, while integration tests are a bit hit-and-miss (mostly because most projects use mocking for these, and the package versions can differ drastically with the one tested upstream).

We could also move these to `test-external` as the `optional-dependency` (those that are optional) and keep the `unit`/`integration`/functional` categorization like this.

Reply:
Thanks for the guidance! Actually I have moved all the tests that involve a non-core dependency to `tests/external`.

All the tests in `unit`, `functional` and `integration` pass, with very little dependencies. I think all these tests should always pass.

I have also included the tests in the package (but I was not sure whether `tests/external` should be included too?)

Reply:
About packaging, yes it is better to be safe and include everything. Also, `sdist` only picks up on git sources (non-git sources are added as artifacts), and the only other one in the list is `demo` which is good to include as well

Reply:
About the tests 👍 to the organization. Just a comment, probably you are aware of the practice of mirroring the source files with the test files, but that is only meaningful in unit-tests, which you seem to be following, so 👍.

There is one confusing part, why areall of the tests as python packages? Pytest should automatically resolve them from the `test_*` naming scheme.

Reply:
Thanks @LecrisUT !

Can you point me to an exemple re the differentiated coverage ?

Re the inits in tests, indeed I am planning to remove them in this PR. Right now I need them because they import from utils.py but I do plan to use fixtures instead of list_notebooks.

Reply:
Just quick links from my projects:
- `.codecov.yaml`: https://github.com/spglib/spglib/blob/b02dedb2eb4b5700ab75c7057f4e7d4d2f38acf8/.codecov.yaml#L14-L31
- Github action: https://github.com/spglib/spglib/blob/b02dedb2eb4b5700ab75c7057f4e7d4d2f38acf8/.github/workflows/step_test.yaml#L103-L109

Reply:
Nice! Thanks for sharing! I'll try this asap.